File: Olympiad/COCI/official/2009/contest2_solutions/svada.cpp
#include <cstdio>

#define MAXN 100

int t, n, m;
int a[MAXN], b[MAXN], c[MAXN], d[MAXN];

void load() {
   scanf( "%d", &t );
   scanf( "%d", &n );
   for( int i = 0; i < n; ++i ) scanf( "%d%d", &a[i], &b[i] );
   scanf( "%d", &m );
   for( int i = 0; i < m; ++i ) scanf( "%d%d", &c[i], &d[i] );
}

int f( int v, int x, int y ) {
   if( v < x ) return 0;
   return (v-x)/y + 1;
}

int koliko( int v, int *x, int *y, int q ) {
   int ret = 0;
   for( int i = 0; i < q; ++i ) ret += f( v, x[i], y[i] );
   return ret;
}

int moze( int v ) {
   return koliko( v, a, b, n ) <= koliko( t-v, c, d, m );
}

int solve() {
   int l = 1, r = t;
   while( r-l > 1 ) {
      int tt = (l+r)/2;
      if( moze(tt) ) l = tt;
      else r = tt;
   }
   return l;
}

int main(void) {
   load();
   printf( "%d\n", solve() );
   return 0;
}

File: Olympiad/COCI/official/2009/contest2_solutions/svada.pas
program svada;

const MAXN = 100;

type
   niz = array[1..MAXN] of longint;

var
   t, n, m : longint;
   a, b, c, d : niz;

procedure load;
var
   i : longint;
begin
   readln( t );
   readln( n );
   for i := 1 to n do readln( a[i], b[i] );
   readln( m );
   for i := 1 to m do readln( c[i], d[i] );
end;

function f( v, x, y : longint ) : longint;
begin
   if v < x then f := 0 else f := (v-x) div y + 1;
end;

function koliko( v, n : longint; var x, y : niz ) : longint;
var
   i : longint;
begin
   koliko := 0;
   for i := 1 to n do koliko := koliko + f( v, x[i], y[i] );
end;

function solve : longint;
var
   l, r, v : longint;
begin
   l := 1;
   r := t;
   while r-l > 1 do begin
      v := (l+r) div 2;
      if koliko( v, n, a, b ) <= koliko( t-v, m, c, d ) then l := v else r := v;      
   end;
   solve := l;
end;

begin
   load;
   writeln( solve() );   
end.
